File: frontend/src/services/notificationService.ts
import api from './api';

export enum NotificationType {
    AUCTION_WON = 'AUCTION_WON',
    PAYMENT_RECEIVED = 'PAYMENT_RECEIVED',
    OUTBID = 'OUTBID',
    AUCTION_ENDED = 'AUCTION_ENDED'
}

export interface Notification {
    id: string;
    recipientId: string;
    message: string;
    type: NotificationType;
    read: boolean;
    createdAt: string;
    relatedEntityId: string;
}

const notificationService = {
    getNotifications: async (): Promise<Notification[]> => {
        const response = await api.get('/notifications');
        return response.data;
    },

    getUnreadNotifications: async (): Promise<Notification[]> => {
        const response = await api.get('/notifications/unread');
        return response.data;
    },

    getUnreadCount: async (): Promise<number> => {
        const response = await api.get('/notifications/unread-count');
        return response.data.count;
    },

    markAsRead: async (id: string): Promise<void> => {
        await api.put(`/notifications/${id}/read`);
    },

    markAllAsRead: async (): Promise<void> => {
        await api.put('/notifications/read-all');
    }
};

export default notificationService;

File: src/main/java/com/springboot_projects/auction_app_api/service/NotificationService.java
package com.springboot_projects.auction_app_api.service;

import com.springboot_projects.auction_app_api.model.Notification;
import com.springboot_projects.auction_app_api.repository.NotificationRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.messaging.simp.SimpMessagingTemplate;
import org.springframework.stereotype.Service;

import java.util.List;

@Service
public class NotificationService {

    @Autowired
    private NotificationRepository notificationRepository;

    @Autowired
    private SimpMessagingTemplate messagingTemplate;

    public Notification createNotification(String recipientId, String message, Notification.NotificationType type,
            String relatedEntityId) {
        Notification notification = new Notification(recipientId, message, type, relatedEntityId);
        Notification savedNotification = notificationRepository.save(notification);

        // Send real-time notification via WebSocket
        messagingTemplate.convertAndSendToUser(recipientId, "/queue/notifications", savedNotification);

        return savedNotification;
    }

    public List<Notification> getUnreadNotifications(String userId) {
        return notificationRepository.findByRecipientIdAndReadOrderByCreatedAtDesc(userId, false);
    }

    public List<Notification> getAllNotifications(String userId) {
        return notificationRepository.findByRecipientIdOrderByCreatedAtDesc(userId);
    }

    public void markAsRead(String notificationId) {
        notificationRepository.findById(notificationId).ifPresent(notification -> {
            notification.setRead(true);
            notificationRepository.save(notification);
        });
    }

    public void markAllAsRead(String userId) {
        List<Notification> unread = notificationRepository.findByRecipientIdAndReadOrderByCreatedAtDesc(userId, false);
        unread.forEach(n -> n.setRead(true));
        notificationRepository.saveAll(unread);
    }

    public long getUnreadCount(String userId) {
        return notificationRepository.countByRecipientIdAndRead(userId, false);
    }
}
